python/vast/utils/arrow.py

```diff
@@ -1,7 +1,72 @@
-import ipaddress
+import ipaddress as ip
+from typing import SupportsBytes
 
 import pyarrow as pa
 
+# Accepts a 128-bit buffer holding an IPv6 address and returns an IPv4 or IPv6
+# address.
+def unpack_ip(buffer: SupportsBytes) -> ip.IPv4Address | ip.IPv6Address:
+    num = int.from_bytes(buffer, byteorder="big")
+    # Convert IPv4 mapped addresses back to regular IPv4.
+    # https://tools.ietf.org/html/rfc4291#section-2.5.5.2
+    if (num >> 32) == 65535:
+        num = num - (65535 << 32)
+    return ip.ip_address(num)
+
+
+class IPAddressScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Address | ip.IPv6Address:
+        return unpack_ip(self.value.as_py())
+
+
+class IPAddressType(pa.ExtensionType):
```

Comment:
In VAST, we so far have called the type just the "address" type, not "IP address."

I understand that IP address is a clearer name, and we'd probably choose that if we were to make a breaking change in VAST, but for now I think it'd be better to align the names for consistency.

```suggestion
class AddressType(pa.ExtensionType):
```